etna/feedback/views.py

```diff
@@ -0,0 +1,197 @@
+import logging
+import uuid
+
+from typing import Any, Dict, Optional
+
+from django.contrib.contenttypes.models import ContentType
+from django.forms import Form
+from django.http import (
+    Http404,
+    HttpRequest,
+    HttpResponse,
+    HttpResponseRedirect,
+    JsonResponse,
+)
+from django.shortcuts import get_object_or_404
+from django.urls import reverse
+from django.utils.decorators import method_decorator
+from django.utils.http import urlencode
+from django.views.decorators.csrf import csrf_exempt
+from django.views.generic import FormView, TemplateView
+
+from wagtail.admin.auth import permission_denied
+from wagtail.admin.views.reports import ReportView
+from wagtail.models import Revision
+
+from etna.feedback.forms import FeedbackForm
+from etna.feedback.models import FeedbackPrompt, FeedbackSubmission
+
+logger = logging.getLogger(__name__)
+
+
+class VersionedFeedbackViewMixin:
+    """
+    Mixin for view classes that are initialized from a prompt_id/version
+    combination, which determine options and messaging shown to the user.
+    """
+
+    def setup(
+        self,
+        request: HttpRequest,
+        prompt_id: uuid.UUID,
+        version: Optional[int] = None,
+        **kwargs,
+    ) -> None:
+        super().setup(request, **kwargs)
+        self.prompt_id = prompt_id
+        self.version = version
+        self.is_ajax = request.POST.get("is_ajax", "false") == "true"
+        if version:
+            self.prompt_revision = get_object_or_404(
+                Revision,
+                content_type=ContentType.objects.get_for_model(FeedbackPrompt),
+                id=version,
+            )
+            self.prompt = self.prompt_revision.as_object()
+            if self.prompt.public_id != prompt_id:
+                raise Http404("Bad prompt_id / version combination.")
+        else:
+            self.prompt = get_object_or_404(
+                FeedbackPrompt.objects.all().select_related("latest_revision"),
+                public_id=prompt_id,
+            )
+            self.prompt_revision = self.prompt.latest_revision
+
+
+@method_decorator(csrf_exempt, name="dispatch")
```

Comment:
Question: Why is CSRF exempt ? Is there an alternative way?

Reply:
Because CSRF is not a risk here, and using CSRF tokens forces the generation of cookies that get in the way of caching and other things. Take a look at the definition here: https://owasp.org/www-community/attacks/csrf